doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,903 @@
+changes needed on different platforms in the common case of an OS network
+socket. (Use of an `int` here is strictly incorrect for Windows; however, this
+style of usage is prevalent in the OpenSSL codebase, so for consistency we
+continue the pattern here.)
+
+For `BIO_s_ssl`, the `BIO_get_[rw]poll_descriptor` functions are equivalent to
+the `SSL_get_[rw]poll_descriptor` functions. The `SSL_get_[rw]poll_descriptor`
+functions are equivalent to calling `BIO_get_[rw]poll_descriptor` on the
+underlying BIOs provided to the SSL object. For a socket BIO, this will likely
+just yield the socket's FD. For memory-based BIOs, see below.
+
+The following APIs return boolean values which indicate whether readability and
+writability are currently desired:
+
+```c
+int SSL_want_net_read(SSL *ssl);
+int SSL_want_net_write(SSL *ssl);
+```
+
+The following API yields a current deadline (if any) after which the QUIC state
+machine should be ticked:
+
+```c
+/* Returns milliseconds or -1 (infinity). */
+int64_t SSL_get_tick_timeout(SSL *ssl);
+```
+
+Finally, the tick function allows the QUIC state machine to be ticked without
+performing any specific operation:
+
+```c
+int SSL_tick(SSL *ssl);
+```
+
+Equivalent BIO methods will be defined for each of the above.
+
+#### Supporting Non-Pollable BIOs
+
+Where we are provided with a non-pollable BIO, we cannot provide the application
+with any primitive used for synchronisation and it is assumed that the
+application will handle its own network I/O, for example via a `BIO_dgram_pair`.
+
+When libssl calls `BIO_get_[rw]poll_descriptor` on the underlying BIO, the call
+fails, indicating that a non-pollable BIO is being used. Thus, if an application
+calls `SSL_get_[rw]poll_descriptor`, that call also fails.
+
+`SSL_get_tick_timeout` is still functional and must be handled by the
+application.
+
+There are various circumstances which need to be handled:
+
+  - The QUIC implementation wants to write data to the network but
+    is currently unable to (e.g. `BIO_dgram_pair` is full).
+
+    This is not hard as our internal TX record layer allows arbitrary buffering.
+    The only limit comes when QUIC flow control (which only applies to
+    application stream data) applies a limit; then we must return
+    `SSL_ERROR_WANT_WRITE`.
+
+  - The QUIC implementation wants to read data from the network
+    but is currently unable to (e.g. `BIO_dgram_pair` is empty).
+
+    Here we need to return `SSL_ERROR_WANT_READ`; we thereby support libssl's
+    classic nonblocking I/O interface.
+
+### Configuration of Blocking vs. Non-Blocking Mode
+
+Traditionally an SSL object has operated either in blocking mode or non-blocking
+mode without requiring explicit configuration; if a socket returns EWOULDBLOCK
+or similar, it is handled appropriately, and if a socket call blocks, there is
+no issue. Since the QUIC implementation is building on non-blocking I/O, this
+implicit configuration of non-blocking mode is not feasible.
```

Comment:
It's not just this that causes a problem. Some responsibilities that were previously at the socket level are now at the libssl level. So things that would block before never get as far as the socket in QUIC (e.g. flow/congestion control). We will need to simulate the blocking to replicate this behaviour.

Reply:
This is the idea, yes.